File: src/main/java/AoC2017_16.java
import static java.util.stream.Collectors.toList;

import com.github.pareronia.aoc.StringOps;
import com.github.pareronia.aoc.StringOps.StringSplit;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.Arrays;
import java.util.List;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2017_16 extends SolutionBase<AoC2017_16.Dance, String, String> {

    private static final String PROGRAMS = "abcdefghijklmnop";

    private AoC2017_16(final boolean debug) {
        super(debug);
    }

    public static AoC2017_16 create() {
        return new AoC2017_16(false);
    }

    public static AoC2017_16 createDebug() {
        return new AoC2017_16(true);
    }

    @Override
    protected Dance parseInput(final List<String> inputs) {
        return Dance.fromInput(inputs.getFirst());
    }

    @Override
    public String solvePart1(final Dance dance) {
        return dance.execute(PROGRAMS, 1);
    }

    @Override
    public String solvePart2(final Dance dance) {
        int cnt = 0;
        String ans = PROGRAMS;
        while (true) {
            ans = dance.execute(ans, 1);
            cnt++;
            if (PROGRAMS.equals(ans)) {
                break;
            }
        }
        return dance.execute(PROGRAMS, 1_000_000_000 % cnt);
    }

    @Override
    protected void samples() {
        final String programs = "abcde";
        assert "cdeab".equals(String.valueOf(Spin.fromInput("3").execute(programs.toCharArray())));
        assert "abced"
                .equals(String.valueOf(Exchange.fromInput("3/4").execute(programs.toCharArray())));
        assert "cbade"
                .equals(String.valueOf(Partner.fromInput("a/c").execute(programs.toCharArray())));
        final AoC2017_16 test = createDebug();
        final Dance dance = test.parseInput(List.of("s1,x3/4,pe/b"));
        assert "baedc".equals(dance.execute(programs, 1));
        assert "ceadb".equals(dance.execute(programs, 2));
    }

    public static void main(final String[] args) throws Exception {
        create().run();
    }

    @FunctionalInterface
    private interface Move {

        @SuppressWarnings("PMD.UseVarargs")
        char[] execute(char[] programs);
    }

    private record Spin(int amount) implements Move {

        public static Spin fromInput(final String input) {
            return new Spin(Integer.parseInt(input));
        }

        @Override
        public char[] execute(final char[] programs) {
            final char[] ans = new char[programs.length];
            System.arraycopy(programs, programs.length - amount, ans, 0, amount);
            System.arraycopy(programs, 0, ans, amount, programs.length - amount);
            return ans;
        }
    }

    private record Exchange(int pos1, int pos2) implements Move {

        public static Exchange fromInput(final String input) {
            final StringSplit sp = StringOps.splitOnce(input, "/");
            return new Exchange(Integer.parseInt(sp.left()), Integer.parseInt(sp.right()));
        }

        @Override
        public char[] execute(final char[] programs) {
            final char tmp = programs[this.pos2];
            programs[this.pos2] = programs[this.pos1];
            programs[this.pos1] = tmp;
            return programs;
        }
    }

    private record Partner(Character program1, Character program2) implements Move {

        public static Partner fromInput(final String input) {
            final StringSplit sp = StringOps.splitOnce(input, "/");
            return new Partner(sp.left().charAt(0), sp.right().charAt(0));
        }

        @Override
        public char[] execute(final char[] programs) {
            int pos1 = -1;
            int pos2 = -1;
            for (int i = 0; i < programs.length; i++) {
                if (programs[i] == this.program1) {
                    pos1 = i;
                }
                if (programs[i] == this.program2) {
                    pos2 = i;
                }
                if (pos1 >= 0 && pos2 >= 0) {
                    break;
                }
            }
            final char tmp = programs[pos2];
            programs[pos2] = programs[pos1];
            programs[pos1] = tmp;
            return programs;
        }
    }

    record Dance(List<Move> moves) {

        public static Dance fromInput(final String input) {
            final List<Move> moves =
                    Arrays.stream(input.split(","))
                            .map(
                                    s ->
                                            switch (s.substring(0, 1)) {
                                                case "s" -> Spin.fromInput(s.substring(1));
                                                case "x" -> Exchange.fromInput(s.substring(1));
                                                default -> Partner.fromInput(s.substring(1));
                                            })
                            .collect(toList());
            return new Dance(moves);
        }

        public String execute(final String programsIn, final int reps) {
            char[] programs = programsIn.toCharArray();
            for (int i = 0; i < reps; i++) {
                for (final Move move : this.moves) {
                    programs = move.execute(programs);
                }
            }
            return String.valueOf(programs);
        }
    }
}

File: src/main/python/AoC2017_16.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 16
#
import sys
from dataclasses import dataclass
from string import ascii_lowercase
from typing import Protocol
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase

TEST = "s1,x3/4,pe/b"

PROGRAMS = ascii_lowercase[:16]


class Move(Protocol):
    def execute(self, d: list[str]) -> list[str]:
        pass


@dataclass(frozen=True)
class Spin(Move):
    amount: int

    @classmethod
    def from_input(cls, s: str) -> Self:
        return cls(int(s))

    def execute(self, d: list[str]) -> list[str]:
        n = len(d) - self.amount
        return d[n:] + d[:n]


@dataclass(frozen=True)
class Exchange(Move):
    pos_1: int
    pos_2: int

    @classmethod
    def from_input(cls, s: str) -> Self:
        a, b = s.split("/")
        return cls(int(a), int(b))

    def execute(self, d: list[str]) -> list[str]:
        tmp = d[self.pos_1]
        d[self.pos_1] = d[self.pos_2]
        d[self.pos_2] = tmp
        return d


@dataclass(frozen=True)
class Partner(Move):
    program_1: str
    program_2: str

    @classmethod
    def from_input(cls, s: str) -> Self:
        a, b = s.split("/")
        return cls(a, b)

    def execute(self, d: list[str]) -> list[str]:
        pos_1, pos_2 = -1, -1
        for i in range(len(d)):
            if d[i] == self.program_1:
                pos_1 = i
            if d[i] == self.program_2:
                pos_2 = i
            if pos_1 >= 0 and pos_2 >= 0:
                break
        tmp = d[pos_1]
        d[pos_1] = d[pos_2]
        d[pos_2] = tmp
        return d


@dataclass(frozen=True)
class Dance:
    moves: list[Move]

    @classmethod
    def parse_move(cls, s: str) -> Move:
        match s[0]:
            case "s":
                return Spin.from_input(s[1:])
            case "x":
                return Exchange.from_input(s[1:])
            case _:
                return Partner.from_input(s[1:])

    @classmethod
    def from_input(cls, inputs: InputData) -> Self:
        return cls([cls.parse_move(s) for s in next(iter(inputs)).split(",")])

    def execute(self, string: str, reps: int) -> str:
        d = list(string)
        for _ in range(reps):
            for move in self.moves:
                d = move.execute(d)
        return "".join(d)


Input = Dance
Output1 = str
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Dance.from_input(input_data)

    def part_1(self, dance: Dance) -> str:
        return dance.execute(PROGRAMS, reps=1)

    def part_2(self, dance: Dance) -> str:
        cnt = 0
        ans = PROGRAMS
        while True:
            ans = dance.execute(ans, reps=1)
            cnt += 1
            if ans == PROGRAMS:
                break
        return dance.execute(PROGRAMS, reps=1_000_000_000 % cnt)

    def samples(self) -> None:
        assert Spin.from_input("3").execute(list("abcde")) == list("cdeab")
        assert Exchange.from_input("3/4").execute(list("abcde")) == list(
            "abced"
        )
        assert Partner.from_input("a/c").execute(list("abcde")) == list(
            "cbade"
        )
        assert Dance.from_input([TEST]).execute("abcde", 1) == "baedc"
        assert Dance.from_input([TEST]).execute("abcde", 2) == "ceadb"


solution = Solution(2017, 16)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
